source/dashboard/main.c

```diff
@@ -433,6 +432,114 @@ void heartBeatLED()
     }
 }
 
+static int8_t encoder_count;
+
+void EXTI9_5_IRQHandler(void) {
+    // Check if EXTI9 triggered the interrupt
+    // PD9: Quad Encoder B
+    if (EXTI->PR & EXTI_PR_PR9) { 
+
+        // Interrupt handling for quad here
+        encoder_ISR();
+        
+        EXTI->PR = EXTI_PR_PR9; // Clear the interrupt pending bit for EXTI9
+
+    }
+}
+
+void EXTI15_10_IRQHandler() {
+    // Check if the interrupt was caused by EXTI11
+    if (EXTI->PR & EXTI_PR_PR11) {
+        PHAL_toggleGPIO(PRCHG_LED_GPIO_Port, PRCHG_LED_Pin);
+        SEND_START_BUTTON(q_tx_can, 1);
+
+        // Clear the interrupt pending bit for EXTI11
+        EXTI->PR |= EXTI_PR_PR11;
+    }
+    
+    // Check if EXTI10 triggered the interrupt
+    // PD9: Quad Encoder A
+    if (EXTI->PR & EXTI_PR_PR10) { 
+
+        // Interrupt handling for quad here
+        encoder_ISR();
+
+        // Clear the interrupt pending bit for EXTI10
+        EXTI->PR |= EXTI_PR_PR10; 
+    }
+}
+
+// Quad Encoder ISR
+void encoder_ISR() {
+    static uint8_t previousA = 0U;
+    static uint8_t previousB = 0U;
+
+    uint8_t currentA = PHAL_readGPIO(ENC_A_GPIO_Port, ENC_A_Pin);
+    uint8_t currentB = PHAL_readGPIO(ENC_B_GPIO_Port, ENC_B_Pin);
+
+    if ((previousA == 0 && previousB == 0) && (currentA == 1 && currentB == 0)) {
+        // Clockwise movement detected
+        encoder_count++;
+    } else if ((previousA == 1 && previousB == 0) && (currentA == 1 && currentB == 1)) {
+        // Clockwise movement detected
+        encoder_count++;
+    } else if ((previousA == 1 && previousB == 1) && (currentA == 0 && currentB == 1)) {
+        // Clockwise movement detected
+        encoder_count++;
+    } else if ((previousA == 0 && previousB == 1) && (currentA == 0 && currentB == 0)) {
+        // Clockwise movement detected
+        encoder_count++;
+    } else if ((previousA == 1 && previousB == 0) && (currentA == 0 && currentB == 0)) {
+        // Counterclockwise movement detected
+        encoder_count--;
+    } else if ((previousA == 1 && previousB == 1) && (currentA == 1 && currentB == 0)) {
+        // Counterclockwise movement detected
+        encoder_count--;
+    } else if ((previousA == 0 && previousB == 1) && (currentA == 1 && currentB == 1)) {
+        // Counterclockwise movement detected
+        encoder_count--;
+    } else if ((previousA == 0 && previousB == 0) && (currentA == 0 && currentB == 1)) {
+        // Counterclockwise movement detected
+        encoder_count--;
+    }
+
+    previousA = currentA;
```

Comment:
Proposed method:

The switch only has 4 possible output states. You could convert from the gray code to the state [0,3].
Store the previous state as the static variable.
Then, on interrupt, decipher the gray code using a switch case to get the new position [0,3].

Update the tick count by the delta in position.